CORE-00 | Update Sampling Documentation for Sampling Refactoring (#4827)
odigos-io/odigos: docs/snippets/shared/pipeline/actions/sampling/head-and-tail-sampling.mdx

## Types of sampling implementation

This page contrasts head sampling (decisions in the **Data Collector** when a trace starts) and tail sampling (decisions at the **Odigos Gateway** after spans are brought together). For how Odigos treats whole traces and root-span decisions, see [Introduction to Sampling](./introduction).

> ### Head sampling (Data Collector)

Reviewer: nit: it's in the Agent code itself, not Data Collector

Author: Corrected

Head sampling is the most basic and efficient form of sampling. The decision is made at the root span before child spans are processed; the trace in its entirety is either kept or dropped to match that decision.

> - **Decision point** — At the root span as the trace starts, inside the Data Collector. The same keep or drop applies to the entire trace (parent-based).

Reviewer: ```suggestion
- **Decision point** — At the root span as the trace starts, inside the supported instrumentation agents. The same keep or drop applies to the entire trace (parent-based).
```

Author: Replaced

> - **Context** — Limited to root-level fields: service name, entry endpoint, and similar. Typical rule inputs are things like service, root span details, and specific routes (e.g. `foo/about`). You cannot base the decision on errors or latency that appear only on child spans.

Reviewer: ```suggestion
- **Context** — Limited to root-level fields: service name, entry endpoint, and similar. Typical rule inputs are things like service, root span details, and specific routes (e.g. `foo/about`). You cannot base the decision on errors or latency that appear only on child spans, or at span end time (http response codes, etc)
```

Author: Replaced

> - **Efficiency** — High for the pipeline: for traces that are dropped, only the root span is processed, and Odigos Gateway work stays low on that path. Volume is cut early, before downstream storage, indexing, and query cost.

Reviewer: If we want to be more accurate, we can say all the spans in a dropped trace are skipped right at span creation time, which uses almost no resources. they are not processed in the agent, not serialized into eBPF maps, not processed at odiglet / data collection, and requires no networking / IO, which is very efficient (when span metrics are enabled this is not the case, but a special case)

Author: Adjusted

**When to use it:**

- **High-Volume Environments:** If your system generates an overwhelming number of traces, head sampling is the most efficient way to drop data early, saving CPU and memory on your application agents and reducing network traffic immediately.

- **Simple Logic:** Use it when you only need to sample based on known "head" attributes, such as the service name, the initial endpoint (e.g., /health vs. /api/orders), or a simple percentage (e.g., "keep 5% of all traffic").

- **Resource Constraints on Collectors:** Since decisions happen at the source, your collectors/gateways don't need to buffer or process data that is already marked to be dropped.

**The Trade-off:** <br/>
It is a "blind" decision. Because the decision is made at the start, you cannot know if that specific trace will eventually encounter an error or experience high latency. You risk losing the exact "needle in the haystack" traces you might need for debugging.

### Tail-Based Sampling

> Tail-based sampling waits until the trace is complete (or a timeout is reached) before deciding whether to keep it. This is typically handled at the collector or gateway level.

Reviewer: ```suggestion
Tail-based sampling waits until the trace is complete (or a timeout is reached) before deciding whether to keep it. This is handled at the gateway collector level.
```

Author: Replaced
